packages/runtime/datastore/src/channelDeltaConnection.ts

```diff
@@ -55,17 +92,41 @@ export class ChannelDeltaConnection implements IDeltaConnection {
 	}
 
 	public reSubmit(content: any, localOpMetadata: unknown) {
-		this.handler.reSubmit(content, localOpMetadata);
+		process(content, localOpMetadata, this.handler.reSubmit.bind(this.handler));
 	}
 
 	public rollback(content: any, localOpMetadata: unknown) {
 		if (this.handler.rollback === undefined) {
 			throw new Error("Handler doesn't support rollback");
 		}
-		this.handler.rollback(content, localOpMetadata);
+		process(content, localOpMetadata, this.handler.rollback.bind(this.handler));
 	}
 
 	public applyStashedOp(content: any): unknown {
-		return this.handler.applyStashedOp(content);
+		try {
+			this.statedOpMd = createStashedOpMetadata();
+			const md = this.handler.applyStashedOp(content);
+			if (md !== undefined) {
+				// temporary assert while we migrate dds to the new pattern
+				// after migration we will always return statedOpMd
```

Comment:
are there obstacles longer-term with removing the scheme implemented here to "coalesce" DDS ops at the channel delta connection layer such that container-runtime only sees a single op? I ask as this still feels like some asymmetry between the resubmit path and the apply stashed op path: in the resubmit case, container-runtime can actually see one op turn into multiple ops whereas with this scheme, from ContainerRuntime's perspective a single applyStashedOp call always results in one submitted op (and one that's kinda bogus, since upon resubmit its contents are pretty much ignored in favor of the actual calls to submit that happened here in `this.handle.applyStashedOp(content)``).

Totally fine with this for now, just wondering where you're imagining things happening longer term here, since if we can simplify the scheme even more that would be great. If not, still better to have the complexity at a shared layer than pawn it off on all the DDSes.

Reply:
apply stashed op is different from resubmit, as we don't know if the op was already sent previously or not, for resubmit we know it hasn't. this means ops that go through apply stashed op always need to be there in case the op already exists on the wire. there is more work coming via pending state, specifically gathering before close, this will mean that resubmit starts to care as well, as it will need to persist and resend ops with ids that can be identified as the same. that isn't designed yet, but it might be easiest to push it down as well, so resubmit always returns a single op to the runtime and has an id, even if inside it has 0 or more ops.